include/jemalloc/internal/cache_bin.h

```diff
@@ -377,7 +410,122 @@ cache_bin_dalloc_easy(cache_bin_t *bin, void *ptr) {
 	return true;
 }
 
-/**
+/* TODO: move to san.c after rebase */
+JEMALLOC_ALWAYS_INLINE void
+cache_bin_junk_ptr_locations(void *ptr, size_t usize, void **first, void **mid,
+    void **last) {
+	size_t ptr_sz = sizeof(void *);
+
+	*first = ptr;
+
+	*mid = (void *)((uintptr_t)ptr + ((usize >> 1) & ~(ptr_sz - 1)));
+	assert(*first != *mid || usize == ptr_sz);
+	assert((uintptr_t)*first <= (uintptr_t)*mid);
+
+	/*
+	 * When usize > 32K, the gap between requested_size and usize is greater
+	 * than 4K -- this means the last write may access an likely-untouched
+	 * page (default settings w/ 4K pages).  However by default the tcache
+	 * only goes up to the 32K size class, and is usually tuned lower
+	 * instead of higher, which makes it less of a concern.
+	 */
+	*last = (void *)((uintptr_t)ptr + usize - sizeof(uintptr_t));
+	assert(*first != *last || usize == ptr_sz);
+	assert(*mid != *last || usize <= ptr_sz * 2);
+	assert((uintptr_t)*mid <= (uintptr_t)*last);
+}
+
+JEMALLOC_ALWAYS_INLINE void
+cache_bin_junk_ptr(cache_bin_t *bin, void *ptr, szind_t binind) {
+	size_t usize = sz_index2size(binind);
+	if (config_debug) {
+		memset(ptr, (uint8_t)uaf_detect_junk, usize);
+		return;
+	}
+
+	void *first, *mid, *last;
+	cache_bin_junk_ptr_locations(ptr, usize, &first, &mid, &last);
+	*(uintptr_t *)first = uaf_detect_junk;
+	*(uintptr_t *)mid = uaf_detect_junk;
+	*(uintptr_t *)last = uaf_detect_junk;
+}
+
+JEMALLOC_ALWAYS_INLINE bool
+cache_bin_stash_ptr(cache_bin_t *bin, void *ptr) {
+	if (cache_bin_full(bin)) {
+		return true;
+	}
+
+	/* Stash at the full position, in the [full, head) range. */
+	uint16_t low_bits_head = (uint16_t)(uintptr_t)bin->stack_head;
+	assert(low_bits_head != bin->low_bits_full);
```

Comment:
Thanks, this is not exactly what I mean though - this assert is equivalent to `!cache_bin_full(bin)`, after you added this function it can be reused here.

Reply:
Ah I see. Will fix.